xds/internal/xdsclient/controller/controller.go

```diff
@@ -0,0 +1,163 @@
+	updateValidator xdsresource.UpdateValidatorFunc
+	logger          *grpclog.PrefixLogger
+
+	cc               *grpc.ClientConn // Connection to the management server.
+	vClient          version.VersionedClient
+	stopRunGoroutine context.CancelFunc
+
+	backoff  func(int) time.Duration
+	streamCh chan grpc.ClientStream
+	sendCh   *buffer.Unbounded
+
+	mu sync.Mutex
+	// Message specific watch infos, protected by the above mutex. These are
+	// written to, after successfully reading from the update channel, and are
+	// read from when recovering from a broken stream to resend the xDS
+	// messages. When the user of this client object cancels a watch call,
+	// these are set to nil. All accesses to the map protected and any value
+	// inside the map should be protected with the above mutex.
+	watchMap map[xdsresource.ResourceType]map[string]bool
+	// versionMap contains the version that was acked (the version in the ack
+	// request that was sent on wire). The key is rType, the value is the
+	// version string, becaues the versions for different resource types should
+	// be independent.
+	versionMap map[xdsresource.ResourceType]string
+	// nonceMap contains the nonce from the most recent received response.
+	nonceMap map[xdsresource.ResourceType]string
+
+	// Changes to map lrsClients and the lrsClient inside the map need to be
+	// protected by lrsMu.
+	//
+	// TODO: after LRS refactoring, each controller should only manage the LRS
+	// stream to its server. LRS streams to other servers should be managed by
+	// other controllers.
+	lrsMu      sync.Mutex
+	lrsClients map[string]*lrsClient
+}
+
+// New creates a new controller.
+func New(config *bootstrap.ServerConfig, updateHandler pubsub.UpdateHandler, validator xdsresource.UpdateValidatorFunc, logger *grpclog.PrefixLogger) (_ *Controller, retErr error) {
+	switch {
+	case config == nil:
+		return nil, errors.New("xds: no xds_server provided")
+	case config.ServerURI == "":
+		return nil, errors.New("xds: no xds_server name provided in options")
+	case config.Creds == nil:
+		return nil, errors.New("xds: no credentials provided in options")
+	case config.NodeProto == nil:
+		return nil, errors.New("xds: no node_proto provided in options")
+	}
+
+	dopts := []grpc.DialOption{
+		config.Creds,
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:    5 * time.Minute,
+			Timeout: 20 * time.Second,
+		}),
+	}
+
+	ret := &Controller{
+		config:          config,
+		updateValidator: validator,
+		updateHandler:   updateHandler,
+
+		backoff:    backoff.DefaultExponential.Backoff, // TODO: should this be configurable?
+		streamCh:   make(chan grpc.ClientStream, 1),
+		sendCh:     buffer.NewUnbounded(),
+		watchMap:   make(map[xdsresource.ResourceType]map[string]bool),
+		versionMap: make(map[xdsresource.ResourceType]string),
+		nonceMap:   make(map[xdsresource.ResourceType]string),
+
+		lrsClients: make(map[string]*lrsClient),
+	}
+
+	defer func() {
+		if retErr != nil {
+			if ret.cc != nil {
```

Comment:
On the other hand, I want to keep this as a reminder to close any resource added in the future.
We were leaking resources in error cases, until the recent refactors I made.

Or should `Close` check for `nil`s? Then this can be `defer Close()`.

Reply:
Changed to defer `ret.Close()`